src/System.Drawing.Common/src/System/Drawing/GdiPlusStreamHelper.Unix.cs

```diff
@@ -41,164 +41,80 @@ namespace System.Drawing
         {
-            if (buf == IntPtr.Zero && peek)
-            {
+            if ((buf == null && peek) || !_stream.CanRead)
                 return -1;
-            }
 
-            if (bufsz > managedBuf.Length)
-                managedBuf = new byte[bufsz];
-            int bytesRead = 0;
-            long streamPosition = 0;
+            if (bufsz <= 0)
+                return 0;
 
-            if (bufsz > 0)
+            int read = 0;
+            long originalPosition = 0;
+            if (peek)
             {
-                if (stream.CanSeek)
-                {
-                    streamPosition = stream.Position;
-                }
-                if (start_buf_len > 0)
-                {
-                    if (start_buf_len > bufsz)
-                    {
-                        Array.Copy(start_buf, start_buf_pos, managedBuf, 0, bufsz);
-                        start_buf_pos += bufsz;
-                        start_buf_len -= bufsz;
-                        bytesRead = bufsz;
-                        bufsz = 0;
-                    }
-                    else
-                    {
-                        // this is easy
-                        Array.Copy(start_buf, start_buf_pos, managedBuf, 0, start_buf_len);
-                        bufsz -= start_buf_len;
-                        bytesRead = start_buf_len;
-                        start_buf_len = 0;
-                    }
-                }
-
-                if (bufsz > 0)
-                {
-                    try
-                    {
-                        bytesRead += stream.Read(managedBuf, bytesRead, bufsz);
-                    }
-                    catch (IOException)
-                    {
-                        return -1;
-                    }
-                }
-
-                if (bytesRead > 0 && buf != IntPtr.Zero)
-                {
-                    Marshal.Copy(managedBuf, 0, (IntPtr)(buf.ToInt64()), bytesRead);
-                }
-
-                if (!stream.CanSeek && (bufsz == 10) && peek)
-                {
-                    // Special 'hack' to support peeking of the type for gdi+ on non-seekable streams
-                }
-
-                if (peek)
-                {
-                    if (stream.CanSeek)
-                    {
-                        // If we are peeking bytes, then go back to original position before peeking
-                        stream.Seek(streamPosition, SeekOrigin.Begin);
-                    }
-                    else
-                    {
-                        throw new NotSupportedException();
-                    }
-                }
+                originalPosition = _stream.Position;
             }
 
-            return bytesRead;
-        }
+            try
+            {
+                // Stream Span API isn't available in 2.0
+#if netcoreapp20
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
```

Comment:
In the original there is a check for `bufsz>managedBuf.,Length` which can then allocate a new array which may be larger than `default_bufsize`, Is bufsz in the new version limited to <=default_bufsize? If not should this rental request max(default_bufsize,bufsz) ?

Reply:
It might be interesting to make the default buffer size relatively small in debug to see if you can get a test failure here.

Reply:
Good catch! Thanks. With max(default, bufsz) I wasn't able to fail obviously, without it I was.

Not worth making it a small value in Debug since this is for netcoreapp20 and we don't run tests against old frameworks.

Reply:
netcoreapp should fail too, presumably with an AV.

Reply:
Ignore me. I'm not thinking straight. :)

Reply:
:smile: